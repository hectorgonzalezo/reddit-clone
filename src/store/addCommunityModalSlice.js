import { createSlice } from '@reduxjs/toolkit';

export const addCommunityModalSlice = createSlice({
  name: 'addCommunityModalVisible',
  initialState: false,
  reducers: {
    toggleAddCommunityModal: (state, action) => {
      return !state
    }
  },
});

export const { toggleAddCommunityModal } = addCommunityModalSlice.actions;

export const selectAddCommunityModalVisibility = (state) => state.addCommunityModalVisible;

export default addCommunityModalSlice.reducer;
