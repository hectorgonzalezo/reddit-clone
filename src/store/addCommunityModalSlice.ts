import { createSlice } from '@reduxjs/toolkit';

export const addCommunityModalSlice = createSlice({
  name: 'addCommunityModalVisible',
  initialState: false,
  reducers: {
    toggleAddCommunityModal: (state) => !state
  },
});

export const { toggleAddCommunityModal } = addCommunityModalSlice.actions;

export const selectAddCommunityModalVisibility = (state: AddCommunityModalState): boolean => state.addCommunityModalVisible;

export default addCommunityModalSlice.reducer;
