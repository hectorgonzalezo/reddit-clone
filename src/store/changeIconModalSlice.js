import { createSlice } from '@reduxjs/toolkit';

export const changeIconModalSlice = createSlice({
  name: 'changeIconModalVisible',
  initialState: false,
  reducers: {
    toggleChangeIconModal: (state, action) => {
      return !state
    }
  },
});

export const { toggleChangeIconModal } = changeIconModalSlice.actions;

export const selectChangeIconModalVisibility = (state) => state.changeIconModalVisible;

export default changeIconModalSlice.reducer;
