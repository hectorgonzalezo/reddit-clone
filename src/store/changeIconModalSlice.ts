import { createSlice } from '@reduxjs/toolkit';
import { ChangeIconModalState } from '../type';

export const changeIconModalSlice = createSlice({
  name: 'changeIconModalVisible',
  initialState: false,
  reducers: {
    toggleChangeIconModal: (state, action) => !state
  },
});

export const { toggleChangeIconModal } = changeIconModalSlice.actions;

export const selectChangeIconModalVisibility = (state: ChangeIconModalState): boolean => state.changeIconModalVisible;

export default changeIconModalSlice.reducer;
