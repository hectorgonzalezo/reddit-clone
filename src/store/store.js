import { configureStore } from '@reduxjs/toolkit';
import userReducer from './userSlice';
import subredditsReducer from './subredditsSlice';
import loginModalReducer from './loginModalSlice';
import currentSubredditReducer from './currentSubredditSlice';

const store = configureStore({
  reducer: {
    user: userReducer,
    subreddits: subredditsReducer,
    loginModalVisible: loginModalReducer,
    currentSubreddit: currentSubredditReducer,
  }
});

export default store;
