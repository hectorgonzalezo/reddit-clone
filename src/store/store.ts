import { configureStore } from '@reduxjs/toolkit';
import userReducer from './userSlice';
import postsReducer from './postsSlice';
import subredditsReducer from './subredditsSlice';
import loginModalReducer from './loginModalSlice';
import changeIconModalReducer from './changeIconModalSlice';
import addCommunityModalReducer from './addCommunityModalSlice';
import currentSubredditReducer from './currentSubredditSlice';

const store = configureStore({
  reducer: {
    user: userReducer,
    posts: postsReducer,
    subreddits: subredditsReducer,
    loginModalVisible: loginModalReducer,
    changeIconModalVisible: changeIconModalReducer,
    addCommunityModalVisible: addCommunityModalReducer,
    currentSubreddit: currentSubredditReducer,
  }
});

export default store;
