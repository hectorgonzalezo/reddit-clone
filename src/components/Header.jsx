import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import logo from '../assets/Reddit_Mark_OnWhite.png';
import logotype from '../assets/Reddit_Logotype_OnWhite.png';
import homeIcon from '../assets/home_icon.svg';
import arrowDownIcon from '../assets/arrow_down_icon.svg';
import userIcon from '../assets/user_icon.svg';
import Button from './Button';
import SearchBar from './SearchBar';
import DropDown from './DropDown';
import { func, bool } from 'prop-types';
import { selectUser } from '../store/userSlice';
import '../styles/headerStyle.scss';

function Header({ signUpFunc, logInFunc, opaque }) {
  // gets user from redux store
  const user = useSelector(selectUser);
  const [userDropdownVisible, setUserDropdownVisible] = useState(false);

  return (
    <header className={opaque ? "opaque" : ""}>
      <div id="logos">
        <a>
          <img src={logo} alt="reddit logo" />
          <img src={logotype} alt="reddit logotype" />
        </a>
      </div>
      <div id="go-to">
        <button type="button">
          <img src={homeIcon} alt="" className="icon" />
          <h1>Home</h1>
          <img src={arrowDownIcon} alt="" className="icon" />
        </button>
      </div>
      <SearchBar />
      <div id="user-area">
        {/* show sign up and log in button if theres no user logged in */}
        {user.username === undefined ? (
          <>
            <Button text="Sign Up" light onClick={signUpFunc} />
            <Button text="Log In" onClick={logInFunc} />
          </>
        ) : (
          <>
            <button
              type="button"
              onClick={() => setUserDropdownVisible((prev) => !prev)}
            >
              <img src={userIcon} alt="" className="icon" />
              <p>{user.username}</p>
              <img src={arrowDownIcon} alt="" className="icon" />
            </button>
            <DropDown
              visible={userDropdownVisible}
              onClickLogout={() => setUserDropdownVisible((prev) => !prev)}
            />
          </>
        )}
      </div>
    </header>
  );
}

Header.defaultProps = {
  signUpFunc: () => {},
  logInFunc: () => {},
  opaque: false,
};

Header.propTypes = {
  signUpFunc: func,
  logInFunc: func,
  opaque: bool,
};

export default Header;
