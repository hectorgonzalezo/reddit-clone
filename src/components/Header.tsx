import React, { useState, useEffect, SyntheticEvent } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useNavigate, Link } from 'react-router-dom';
import logo from '../assets/Reddit_Mark_OnWhite.png';
import logotype from '../assets/Reddit_Logotype_OnWhite.png';
import homeIcon from '../assets/home_icon.svg';
import arrowDownIcon from '../assets/arrow_down_icon.svg';
import Button from './Button';
import SearchBar from './SearchBar';
import AccountDropDown from './AccountDropDown';
import SubredditsDropDown from './subreddit/SubredditsDropDown';
import { selectUser } from '../store/userSlice';
import { selectCurrentSubreddit, changeCurrentSubreddit } from '../store/currentSubredditSlice';
import { getSubreddit } from '../api/communities';

import '../styles/headerStyle.scss';
import { selectSubreddits } from '../store/subredditsSlice';

interface HeaderProps {
  signUpFunc: () => void,
  logInFunc: () => void,
  opaque: boolean,
};

function Header({
  signUpFunc = () => {},
  logInFunc = () => {},
  opaque = false,
}: HeaderProps): JSX.Element {
  // gets user from redux store
  const user = useSelector(selectUser);
  const subreddits = useSelector(selectSubreddits);
  const currentSubreddit = useSelector(selectCurrentSubreddit);
  const [userDropdownVisible, setUserDropdownVisible] = useState(false);
  const [subredditDropdownVisible, setSubredditDropdownVisible] =
    useState(false);
  const [subredditIcon, setSubredditIcon] = useState(homeIcon);
  const navigate = useNavigate();
  const dispatch = useDispatch();

  function toggleSubredditDropdown(): void {
    setSubredditDropdownVisible((prev) => !prev);
  }

  async function navigateToSubreddit(e: SyntheticEvent): Promise<void> {
    const target = e.target as HTMLAnchorElement;
    const subredditName = target.getAttribute("data-name") as string;

    dispatch(changeCurrentSubreddit(subredditName));
    navigate(`/r/${subredditName}`);
  }

  // closes subreddits and user dropdowns when mouse leaves the header area
  function closeAllDropDowns(): void {
    if (userDropdownVisible) {
      setUserDropdownVisible(false);
    }
    if (subredditDropdownVisible) {
      setSubredditDropdownVisible(false);
    }
  }

  useEffect(() => {
    // change subreddit icon
    if (currentSubreddit !== null) {
      const currentSubredditId = subreddits[currentSubreddit]._id.toString();
      getSubreddit(currentSubredditId)
        .then((data) => {
          if (data.community.icon !== undefined) {
            setSubredditIcon(data.community.icon);
          }
        })
        .catch((error) => console.log(error));
    } else {
      setSubredditIcon(homeIcon);
    }
  }, [currentSubreddit]);

  return (
    <header className={opaque ? "opaque" : ""} onMouseLeave={closeAllDropDowns}>
      <div id="logos">
        <Link to="/">
          <img src={logo} alt="reddit" />
          <img src={logotype} alt="reddit" />
        </Link>
      </div>
      {user.username !== undefined ? (
        <div id="go-to" data-testid="go-to">
          <button
            type="button"
            className="button-show-drop-down"
            onClick={toggleSubredditDropdown}
          >
            <img
              src={subredditIcon}
              alt="chosen subreddit"
              className={subredditIcon === homeIcon ? "icon" : "user-icon"}
            />
            <h1>{currentSubreddit === null ? "Home" : currentSubreddit}</h1>
            <img src={arrowDownIcon} className="icon" alt="" aria-hidden="true" />
          </button>
          <SubredditsDropDown
            dropdownVisible={subredditDropdownVisible}
            toggleDropdown={toggleSubredditDropdown}
            chooseFromDropdown={navigateToSubreddit}
            header
          />
        </div>
      ) : null}
      <SearchBar />
      <div id="user-area" data-testid="user-area">
        {/* show sign up and log in button if theres no user logged in */}
        {user.username === undefined ? (
          <>
            <Button text="Sign Up" light onClick={signUpFunc} />
            <Button text="Log In" onClick={logInFunc} />
          </>
        ) : (
          <>
            <button
              type="button"
              className="button-show-drop-down"
              onClick={() => setUserDropdownVisible((prev) => !prev)}
            >
              <img src={user.icon} alt="user icon" className="icon user-icon" />
              <p>{user.username}</p>
              <img src={arrowDownIcon} alt="" className="icon" />
            </button>
            <AccountDropDown
              userIcon={user.icon !== undefined ? user.icon : ''}
              visible={userDropdownVisible}
              closeFunc={() => setUserDropdownVisible(false)}
            />
          </>
        )}
      </div>
    </header>
  );
}


export default Header;
