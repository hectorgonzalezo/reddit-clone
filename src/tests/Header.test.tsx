import React from 'react';
import { render, screen, container, act, getByRole, getAllByRole } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import Header from '../components/Header';
import store from '../store/store';

jest.mock('../firebase/firebase');

const icon = 'https://firebasestorage.googleapis.com/v0/b/reddit-clone-83ce9.appspot.com/o/user_icon.svg?alt=media&token=50e7a9f1-8508-4d51-aac8-4d1ed9dad7a1';

describe('Sign up modal', () => {
  test('User area should display log in and sign up buttons by default', () => {
    render(<Provider store={store}>
      <BrowserRouter>
        <Header />
      </BrowserRouter>
     </Provider>);

    const userArea = screen.queryByTestId('user-area');
    expect(userArea.children.length).toBe(2);

    const logInButton = screen.getByRole('button', { name: 'Log In' });
    expect(logInButton).toBeInTheDocument();
    const signUpButton = screen.getByRole('button', { name: 'Sign Up' });
    expect(signUpButton).toBeInTheDocument();
    
    expect(container).toMatchSnapshot();
  });

  test('Log in and sign up button call function that opens their modals', () => {
    const mockLogInFunc = jest.fn();
    const mockSignUpFunc = jest.fn();
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Header logInFunc={mockLogInFunc} signUpFunc={mockSignUpFunc} />
        </BrowserRouter>
      </Provider>
    );

    const logInButton = screen.getByRole('button', { name: 'Log In' });
    userEvent.click(logInButton);

    const signUpButton = screen.getByRole('button', { name: 'Sign Up' });
    userEvent.click(signUpButton);

    expect(mockLogInFunc).toBeCalled();
    expect(mockSignUpFunc).toBeCalled();
  });

  test('If given a user, user area should contain a single button with the user info', async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </Provider>
    );

    await act(async () => {
      store.dispatch({
        type: "user/addUser",
        payload: { username: "juan", email: "mock@mock.com", icon, subreddits: []},
      });
    });

    // User name and icon should be displayed
    expect(screen.getByText('juan')).toBeInTheDocument();
    expect(screen.getByAltText('user icon')).toBeInTheDocument();

  });

  test('If given a user, drop down menu should be invisible and shown only when clicking', async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </Provider>
    );

    await act(async () => {
      store.dispatch({
        type: "user/addUser",
        payload: { username: "juan", email: "mock@mock.com", icon, subreddits: []},
      });
    });

    const userArea = screen.queryByTestId('user-area');

    // drop down shoul be invisible
    const dropDown = screen.getByTestId('user-dropdown');
    expect(dropDown).toHaveStyle('display: none');

    // User name and icon should be displayed
    const userButton = getByRole(userArea, 'button');
    userEvent.click(userButton);

    expect(dropDown).not.toHaveStyle('display: none');
  });

  test('Pressing logout on user dropdown logs user out', async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </Provider>
    );

    await act(async () => {
      store.dispatch({
        type: "user/addUser",
        payload: { username: "juan", email: "mock@mock.com", icon, subreddits: [] },
      });
    });

    const userArea = screen.queryByTestId('user-area');

    const dropDown = screen.getByTestId('user-dropdown');
    expect(dropDown).toHaveStyle('display: none');

    const userButton = getByRole(userArea, 'button');
    userEvent.click(userButton);

    const logOutLink = screen.getByTestId('logout-link');

    await act(async () => userEvent.click(logOutLink));


    // there should be no user information
    expect(screen.queryByText('juan')).not.toBeInTheDocument();
    expect(screen.queryByAltText('user icon')).not.toBeInTheDocument();
  });

  test('If given a user, there should be a go to bar', async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </Provider>
    );

    // It shouldn't be there by default
    expect(screen.queryByTestId('go-to')).not.toBeInTheDocument()

    await act(async () => {
      store.dispatch({
        type: "user/addUser",
        payload: { username: "juan", email: "mock@mock.com", icon, subreddits: [] },
      });
    });

    expect(screen.queryByTestId('go-to')).toBeInTheDocument()

  });

  test('Pressing on go to button shows subreddits dropdown', async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </Provider>
    );

    // It shouldn't be there by default

    await act(async () => {
      store.dispatch({
        type: "user/addUser",
        payload: { username: "juan", email: "mock@mock.com", icon, subreddits: [] },
      });
    });

    const goToArea = screen.getByTestId('go-to')
    const goToButton = getByRole(goToArea, 'button');
    const dropDown = screen.getByTestId('subreddits-dropdown');

    expect(dropDown).toHaveStyle('display: none');

    userEvent.click(goToButton);

    expect(dropDown).not.toHaveStyle('display: none');
  });

  test('Selecting a subreddit shows its name and icon in go to bar', async () => {
    const mockSubreddit = {
      name: 'aww',
      dateCreated: (new Date()).toString(),
      description:
          'Things that make you go AWW! Like puppies, bunnies, babies, and so on... A place for really cute pictures and videos!',
      icon,
      members: 0,
      postQuantity: 3,
      subtitle: 'A subreddit for cute and cuddly pictures',
    };

    await act(async () => {
      store.dispatch({
        type: "user/addUser",
        payload: { username: "juan", email: "mock@mock.com", icon, subreddits: [] },
      });
    });

    await act(async () => {
      store.dispatch({
        type: "subreddits/addSubreddit",
        payload: [mockSubreddit],
      });
    });

    await act(async () => {
      store.dispatch({
        type: "currentSubreddit/changeCurrentSubreddit",
        payload: 'aww',
      });
    });

    await act(async () => {
      render(
      <Provider store={store}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </Provider>
    )
    });

    // It shouldn't be there by default

    const goToArea = screen.getByTestId('go-to')
    const goToButton = getByRole(goToArea, 'button');
    const subredditName = getByRole(goToButton, 'heading');
    const subredditIcon = getByRole(goToButton, 'img');

    expect(subredditName).toHaveTextContent('aww');
    expect(subredditIcon).toHaveAttribute('src', icon);
  });
});