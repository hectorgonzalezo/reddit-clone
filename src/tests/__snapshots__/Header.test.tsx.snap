// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign up modal User area should display log in and sign up buttons by default 1`] = `undefined`;

exports[`Sign up modal without user User area should display log in and sign up buttons by default 1`] = `undefined`;
